packages/plugin-response-cache/src/ApolloServerPluginResponseCache.ts

```diff
@@ -0,0 +1,331 @@
+import type {
+  ApolloServerPlugin,
+  BaseContext,
+  CacheHint,
+  GraphQLRequestContext,
+  GraphQLRequestListener,
+  GraphQLResponse,
+} from '@apollo/server';
+import { createHash } from '@apollo/utils.createhash';
+import {
+  KeyValueCache,
+  PrefixingKeyValueCache,
+} from '@apollo/utils.keyvaluecache';
+
+type ValueOrPromise<T> = T | Promise<T>;
+interface Options<TContext = Record<string, any>> {
+  // Underlying cache used to save results. All writes will be under keys that
+  // start with 'fqc:' and are followed by a fixed-size cryptographic hash of a
+  // JSON object with keys representing the query document, operation name,
+  // variables, and other keys derived from the sessionId and extraCacheKeyData
+  // hooks. If not provided, use the cache in the GraphQLRequestContext instead
+  // (ie, the cache passed to the ApolloServer constructor).
+  cache?: KeyValueCache;
+
+  // Define this hook if you're setting any cache hints with scope PRIVATE.
+  // This should return a session ID if the user is "logged in", or null if
+  // there is no "logged in" user.
+  //
+  // If a cacheable response has any PRIVATE nodes, then:
+  // - If this hook is not defined, a warning will be logged and it will not be cached.
+  // - Else if this hook returns null, it will not be cached.
+  // - Else it will be cached under a cache key tagged with the session ID and
+  //   mode "private".
+  //
+  // If a cacheable response has no PRIVATE nodes, then:
+  // - If this hook is not defined or returns null, it will be cached under a cache
+  //   key tagged with the mode "no session".
+  // - Else it will be cached under a cache key tagged with the mode
+  //   "authenticated public".
+  //
+  // When reading from the cache:
+  // - If this hook is not defined or returns null, look in the cache under a cache
+  //   key tagged with the mode "no session".
+  // - Else look in the cache under a cache key tagged with the session ID and the
+  //   mode "private". If no response is found in the cache, then look under a cache
+  //   key tagged with the mode "authenticated public".
+  //
+  // This allows the cache to provide different "public" results to anonymous
+  // users and logged in users ("no session" vs "authenticated public").
+  //
+  // A common implementation of this hook would be to look in
+  // requestContext.request.http.headers for a specific authentication header or
+  // cookie.
+  //
+  // This hook may return a promise because, for example, you might need to
+  // validate a cookie against an external service.
+  sessionId?(
+    requestContext: GraphQLRequestContext<TContext>,
+  ): ValueOrPromise<string | null>;
```

Comment:
we could consider making these all return Promise (ie always async) like we changed the plugin API to do in AS3.  (and if we consistently use `await` rather than `then`/`catch` on them, we don't actually break non-TS folks who accidentally write a sync function)

Reply:
I like that suggestion: [b0bc088](https://github.com/apollographql/apollo-server/pull/6610/commits/b0bc0880f22626d36c201515ab90e37a1d39418b)